Website fixes, limit TCP SYN, separate JSON data from mutexes

deputinizer opened the pull request:
- Mutexes don't need a pointer like `mux *sync.Mutex`
- Status structs separated from service structs
- Removed blocking code that slowed down workers (sleep+mutex)
- Limited TCP SYN (very useful for OVH machines)

deputinizer committed:
fix: proper mutex usage (remove pointers)

deputinizer committed:
fix: separate data from services (JSON from mutexes)

deputinizer committed:
fix: do not connect to 0.0.0.0
Because ISPs block some .ru sites

deputinizer committed:
feat: removed sleep and mutex that was blocking workers
Now pause is implemented in async way i.e.
worker just ignores paused websites.

deputinizer committed:
feat: limit TCP SYN packets per second
Very useful for routers that can't handle 3000 packets/s
Also many ISPs (including OVH) detect SYN flood as an attack (of course)

deputinizer committed:
Merge branch 'main' into tor-socks-fixes

deputinizer committed:
refactor: rename custom fasthttp dialer/resolver

deputinizer committed:
refactor: rename custom TCP dialer

baylanger commented:
Thank you!

baylanger commented:
I am testing and seeing "too fast TCP SYN (dial spam)" - this is expected side effect, right?

deputinizer commented:
> I am testing and seeing "too fast TCP SYN (dial spam)" - this is expected side effect, right?

Yes, it is intentional.
Also you can also lower `Concurrency:      0,` in `TCPDialer`.

At least now it doesn't eat CPU that much.

deputinizer committed:
feat: add proxy chain parser

deputinizer committed:
feat: TCP dial concurrency flag

baylanger commented:
Currently running latest version w/ new options - will report later, so far no issue with your changes.

deputinizer committed:
refactor: organize resolvers

deputinizer committed:
fix: check non-public addr also in fasthttp dialer
Because some domains point to CNAME domains and the first check
in websites.go sometimes doesn't catch it

deputinizer committed:
refactor: cleanup DNS message

deputinizer commented:
Ok, im done for today. 
Let's do all HTTP resolving through Russian DNS in Golang.
In my country ISP's DNS returns 0.0.0.0 so we should make them hardcoded:
```golang
	"194.54.14.186:53":  {},
	"194.54.14.187:53":  {},
	"194.67.7.1:53":     {},
	"194.67.2.109:53":   {},
	"84.252.147.118:53": {},
	"84.252.147.119:53": {},
```

deputinizer committed:
fix: SIGSEGV customtcpdial.go:194

baylanger commented:
Where are the 2 84.* going?

Below current .by & .ru root servers.

Russia
```
$ root_ns=ru && for dns in $(dig ns $root_ns|grep ^$root_ns\.|awk '{print $NF}'); do echo -n "$dns "; host $dns|grep "has address"|grep -E -o "[^^][0-9]{1,3}.[0-9]{1,3}.[0-9]{1,3}.[0-9]{1,3}"|sed -e 's/ //g'; done
a.dns.ripn.net. 193.232.128.6
b.dns.ripn.net. 194.85.252.62
d.dns.ripn.net. 194.190.124.17
e.dns.ripn.net. 193.232.142.17
f.dns.ripn.net. 193.232.156.17
```

Belarus
```
% root_ns=by && for dns in $(dig ns $root_ns|grep ^$root_ns\.|awk '{print $NF}'); do echo -n "$dns "; host $dns|grep "has address"|grep -E -o "[^^][0-9]{1,3}.[0-9]{1,3}.[0-9]{1,3}.[0-9]{1,3}"|sed -e 's/ //g'; done
dns1.tld.becloudby.com. 93.125.25.72
dns2.tld.becloudby.com. 93.125.25.73
dns3.tld.becloudby.com. 185.98.83.4
dns4.tld.becloudby.com. 184.72.17.94
dns5.tld.becloudby.com. 54.180.35.203
```

[edited] ^

deputinizer committed:
fix: ignore too fast dial TCP SYN errors

deputinizer committed:
feat: extra error info about which address timed out

deputinizer commented:
Extra timeout error looks like:
```json
        "https://www.gazprombank.ru": {
            "requests": 2633,
            "errors": 59,
            "last_error_msg": "CustomTCPDialer: 82.202.190.10:443 timed out",
        }
```

deputinizer commented:
> Where are the 2 84.* going?

Don't know 😜

deputinizer committed:
fix: better error tracking

erkexzcx commented:
@deputinizer you are doing an outstanding job.

erkexzcx commented on internal/stoppropaganda/http.go:
```diff
@@ -51,19 +54,19 @@ func fasthttpStatusResponseHandler(ctx *fasthttp.RequestCtx) {
-			dosPausedFor := time.Since(tmpWebsite.dnsLastChecked)
-			if tmpWebsite.paused {
+			dosPausedFor := time.Since(ws.dnsLastChecked)
+			if ws.paused {
```
Bool values aren't thread safe. You need mutex for them.

erkexzcx commented on internal/stoppropaganda/websites.go:
```diff
@@ -400,114 +432,108 @@ func startWebsites() {
+		}
+		website.ValidateDNS()
+		// Check again after DNS validated
+		if website.paused {
```
Bool values aren't thread safe. You need mutex for them.

erkexzcx replied:
There are sync.RWMutex if you are performing lots of READs and only a few WRITEs to the variable.

erkexzcx commented on internal/stoppropaganda/websites.go:
```diff
@@ -400,114 +432,108 @@ func startWebsites() {
-			}
-
+		if !website.ShouldRun() {
+			continue
```
Why don't we sleep for e.g. 1 second here? Without sleeping even for 100ms we are literally wasting CPU.

erkexzcx commented on internal/stoppropaganda/websites.go:
```diff
@@ -400,114 +432,108 @@ func startWebsites() {
+		website.ValidateDNS()
+		// Check again after DNS validated
+		if website.paused {
+			continue
```
Why don't we sleep for e.g. 1 second here? Without sleeping even for 100ms we are literally wasting CPU.

deputinizer replied:
If website is paused, then continue and fetch next website from the channel. 
I think CPU would be wasted only if all of the websites are off.

erkexzcx replied:
> then continue and fetch next website

Yes, you are right. I was consfused with previous logic (this would be bad if we used per-website workers).

erkexzcx commented on internal/stoppropaganda/websites.go:
```diff
@@ -343,13 +345,43 @@ type Website struct {
 	Counter_code500 uint `json:"status_500"`
+}
+
+func (ws *WebsiteStatus) IncreateCounters(responseCode int) {
```
You meant `increase` rather than `increate`? If so, CTRL+F `increate`.

erkexzcx commented on internal/stoppropaganda/stoppropaganda.go:
```diff
@@ -19,18 +20,22 @@ import (
+	flagDNSWorkers  = fs.Int("dnsworkers", 100, "DOS each DNS server with this amount of workers")
+	flagDNSTimeout  = fs.Duration("dnstimeout", time.Second, "timeout of DNS request")
+	flagUserAgent   = fs.String("useragent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36", "User agent used in HTTP requests")
+	dialsPerSecond  = fs.Int("dialspersecond", 2500, "maximum amount of TCP SYN packets sent per second from fasthttp")
```
Please keep naming convention consistant. It should be `flagDialsPerSecond`. Same in next line.

erkexzcx commented on internal/stoppropaganda/resolvefix/resolvefix.go:
```diff
@@ -0,0 +1,36 @@
+	return nil
+}
+
+func ContainsNonPublicIP(ips []net.IP) bool {
```
Unused function. Remove it.

erkexzcx commented on internal/stoppropaganda/resolvefix/resolvefix.go:
```diff
@@ -0,0 +1,36 @@
+	"net"
+)
+
+func CheckNonPublicIPAddrs(addrs []net.TCPAddr) error {
```
Rename to `CheckNonPublicTCPEndpoints` because current naming is confusing

deputinizer committed:
fix: variable/function names

deputinizer committed:
fix: paused mutex
Maybe there should be atomic boolean

erkexzcx commented on internal/stoppropaganda/http.go:
```diff
@@ -16,18 +16,21 @@ func fasthttpRequestHandler(ctx *fasthttp.RequestCtx) {
+	DNS      map[string]*DNSServerStatus `json:"DNS"`
+	Websites map[string]*WebsiteStatus   `json:"Websites"`
+}
+type StatusService struct {
```
Why did you create `StatusService`? In order to wrap between brackets like this?
```json
{
    "DNS": {
        "..."
    },
    "Websites": {
        "..."
    }
{
```

deputinizer replied:
Because after i removed pointer `mux *sync.Mutex` there was a warning on `json.MarshalIndent(statusStruct)`.
`call of json.MarshalIndent copies lock value: StatusStruct contains sync.Mutex`

Maybe mutex pointer was better then...

deputinizer replied:
Also, separation of data and business logic is a good practice :yum:

erkexzcx replied:
This was the reason why I used mutex pointers, so I can copy a structure with mutexes ;) Your approach is also fine.

deputinizer committed:
fix: mistake - negate boolean shouldRun

deputinizer committed:
feat: use unpauseTime instead of dnsLastChecked for countdown
looks the same:
"status": "Paused: Non public IP detected: 0.0.0.0, DOS paused for 4m30.545996828s",

erkexzcx committed:
fix: variables naming

erkexzcx committed:
add: documentation about dialspersecond/dialconcurrency

erkexzcx commented:
Can you make `all_proxy` passable via arguments, something like this?
```go
flagProxy = fs.String("proxy", "socks5://tor:9050", "proxy (or comma separated chain of proxies) used for websites DOS")
```

EDIT: Never mind, I've done it for you.

erkexzcx commented:
Let me know if this looks good to you now and I will merge. Also check if instructions in README.md looks clear to you, so not-so-technical people can use proxy.

erkexzcx committed:
feat: set proxy using standard configuration rather than env variables

erkexzcx committed:
fix: go linter warnings

erkexzcx commented on internal/stoppropaganda/websites.go:
```diff
@@ -400,114 +432,117 @@ func startWebsites() {
+		if !paused {
+			ws.paused = false
+		}
+		shouldRun = !paused
```
You are not doing anything with this variable

deputinizer replied:
`shouldRun` is used as a return value

erkexzcx commented:
Something is not working:
```
        "http://elkursk.ru": {
            "requests": 12173,
            "errors": 12173,
            "last_error_msg": "httpClient.Do: lookup elkursk.ru: Temporary failure in name resolution",
            "status": "Running",
            "status_100": 0,
            "status_200": 0,
            "status_300": 0,
            "status_400": 0,
            "status_500": 0
        },
```

Message should start with `Paused: ` and counters shouldn't increase (in this case for a second).

EDIT: https://github.com/erkexzcx/stoppropaganda/pull/93/commits/14fe334441f4b9d68470a3fb167c975831346952 does not fix it. Still broken somewhere.

erkexzcx committed:
refactor: improve error handling logic

erkexzcx committed:
refactor: simplify logic

erkexzcx committed:
fix: error handling, simplify logic, code style

erkexzcx commented:
Everything should be fixed by now. Working to me. Looks good to you?

erkexzcx commented on internal/stoppropaganda/websites.go:
```diff
@@ -343,13 +345,39 @@ type Website struct {
+}
+func (ws *WebsiteStatus) IncreaseCountersErr(errMsg string) {
+	ws.Requests++
+	if !strings.Contains(errMsg, customtcpdial.ErrTooFastDialSpam.Error()) {
```
Don't you think this will be confusing for users?
```
        "https://mail.ru": {
            "requests": 2911,
            "errors": 15,
            "last_error_msg": "httpClient.Do: timeout",
            "status": "Running",
            "status_100": 0,
            "status_200": 0,
            "status_300": 0,
            "status_400": 0,
            "status_500": 0
        },
```
Should we use `client_errors` and `server_errors` instead, so counters match?

erkexzcx replied:
Fixed with https://github.com/erkexzcx/stoppropaganda/pull/93/commits/e552a85487eb1bfb6749677da2ccd78df2dfd3c0

erkexzcx committed:
refactor: simplify code

erkexzcx commented:
Should be good. Raise issue or another MR if you find any issues.

erkexzcx committed:
fix: status counters

deputinizer commented:
> Should we use client_errors and server_errors instead, so counters match?

I think so

> Don't you think this will be confusing for users?

Yes, it's confusing, I have to make this TCP SYN limit less noticable :/
